File: services/order-service-python/src/domain/services/order_placement_service.py
from typing import List
from ..domain.models.order import Order, OrderId, OrderItem, InvalidOrderException
from ..domain.ports.order_repository import OrderRepository

class OrderPlacementService:
    def __init__(self, order_repository: OrderRepository):
        self.order_repository = order_repository

    def place_order(self, customer_id, items: List[OrderItem]) -> Order:
        # 1. Semantic Domain Validation
        if not items:
            raise InvalidOrderException("Order must have at least one item")

        # 2. Create Domain Entity
        order = Order(
            id=OrderId.generate(),
            customer_id=customer_id,
            items=items
        )

        # 3. Persist via Port
        return self.order_repository.save(order)

File: services/order-service-java/src/main/java/com/example/orderservice/domain/services/OrderPlacementService.java
package com.example.orderservice.domain.services;

import com.example.orderservice.domain.models.*;
import com.example.orderservice.domain.ports.OrderRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class OrderPlacementService {
    private final OrderRepository orderRepository;

    /**
     * Core business logic for placing an order.
     * Pure domain logic: no DTOs, no HTTP, no framework dependencies.
     */
    public Order placeOrder(UUID customerId, List<OrderItem> items) {
        // 1. Semantic Domain Validation
        if (items == null || items.isEmpty()) {
            throw new InvalidOrderException("Order must have at least one item");
        }

        // 2. Create Domain Entity (Logic inside Order.create)
        Order order = Order.create(customerId, items);

        // 3. Persist via Port
        return orderRepository.save(order);
    }
}
